File: decorator/kotlin/FileDataSource.kt
package Decorator

// Temel işlemlerin tanımlandığı IDataSource arayüzünü uygulayan sınıftır.
// UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
internal class FileDataSource(override val fileName: String) : IDataSource {

    override fun readData() {
        println("$fileName readed.")
    }

    override fun writeData(data: Any) {
        println("data was written to $fileName")
    }
}

File: decorator/python/file_data_source.py
from data_source import DataSource


class FileDataSource(DataSource):
    """
    Temel işlemlerin tanımlandığı IDataSource arayüzünü uygulayan sınıftır.
    UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
    """

    def __init__(self, file_name):
        self.file_name = file_name

    def get_file_name(self):
        return self.file_name

    def write_data(self, data):
        print(f"data was written to {self.file_name}.")

    def read_data(self):
        print(f"{self.file_name} readed.")
